miio/device.py

```diff
@@ -246,14 +249,26 @@ def status(self) -> DeviceStatus:
         """Return device status."""
         raise NotImplementedError()
 
+    def cached_status(self) -> DeviceStatus:
+        """Return device status from cache."""
+        if self._status is None:
+            self._status = self.status()
+
+        return self._status
+
     def actions(self) -> Dict[str, ActionDescriptor]:
         """Return device actions."""
         return {}
 
     def settings(self) -> Dict[str, SettingDescriptor]:
         """Return device settings."""
-        settings = self.status().settings()
-        for setting in settings.values():
+        if self._settings is not None:
+            return self._settings
+
+        self._settings = (
+            self.cached_status().settings()
+        )  # NOTE that this already does IO so schould be run in executer job in HA
+        for setting in self._settings.values():
```

Comment:
Instead of handling the initialization inside the methods returning the descriptors, maybe there should be a separate `_initialize_descriptors` that handles the initialization, leaving these getters just to return the existing, cached values?

Calling the initialization procedure could be done, e.g., as part of the initial `info()` request after the model is known?

Reply:
Sounds very good, I implemented it that way!